docs/general/builtins/aggregation.rst

```diff
@@ -832,15 +832,19 @@ Example::
 
 .. _aggregation-topk:
 
-``topk(column, k)``
+``topk(column, k, capacity)``
 -------------------
 
 The ``topk`` aggregate function computes an approximation of the most frequent
 values and their frequencies of a specific column. The result is an array of
 objects containing the most common values and their frequencies ordered by
 frequency. The k parameter is an optional parameter which defines how many
 elements should be returned which defaults to 8. The maximum limit for k is
-10000.
+5000. The capacity parameter is optional and describes the maximum number
+of tracked items,must be in be a power of 2 and defaults to 8192.
```

Comment:
That was in the previous implementation. If you look at a larger power of two number the value from the user to the actual can differ a lot which is confusing.

Reply:
I think the implications of changing the value from a user perspective aren't very clear anyway. However, needing to do a "calculation" to use a feature seems a bit odd. `hyperloglog_distinct`, as far as I know, has the `precision` option, which allows users to [specify the exponent](https://cratedb.com/docs/crate/reference/en/latest/general/builtins/aggregation.html#hyperloglog-distinct-column-precision).

> If you look at a larger power of two number, the value from the user to the actual can differ a lot, which is confusing.

Not fully following. The difference could be max the provided capacity (e.g., capacity=4097, diff=4095), but since the concrete implications are unknown, I think this isn't so important. A user might know how many distinct values a column has and should be able to provide that value. Then again, the question arises if we can't use column statistics to get a better estimate to start with. 🤷

Reply:
2^13 = 8192 and 2^14 = 16384. Now think if a user choose 10000. Since this is the number of tracked items which has a direct impact on the error rate and performance it think it's better to explicit.

Reply:
Column statistics won't help, this is depending on what you want to do. Did you look at some top-k implementations in other systems ?

Reply:
> 2^13 = 8192 and 2^14 = 16384. Now think if a user choose 10000. Since this is the number of tracked items which has a direct impact on the error rate and performance it think it's better to explicit.

But if a user chooses 10000, this implies they need to use 16384 anyway?

> Did you look at some top-k implementations in other systems ?

ClickHouse is using a multiple of `k` 
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/topk
https://clickhouse.com/docs/en/sql-reference/aggregate-functions/reference/approxtopk

which defaults to `k * 3`

Reply:
The previous implementation was similar and when I benchmarked more datasets I found it did not perform well on some data so i made the parameters explicit like in spark.

https://github.com/crate/crate/pull/16646
https://docs.databricks.com/en/sql/language-manual/functions/approx_top_k.html

Reply:
I'd also think that the capacity should grow in a somewhat reasonable way with the `k` parameter.
Otherwise you're kinda forced to provide it, no?

Reply:
The main takeaway from the result here is that the initial capacity should be much larger by default.

Reply:
Spark starts with limit 5 and capacity 10.000 as default value.

Reply:
Now, the maximum for limit is 5000 with a capacity of 8192.

Reply:
> The main takeaway from the result here is that the initial capacity should be much larger by default.

Isn't that trading memory for performance? Running a benchmark where you don't look at the memory consumption and get lower response times doesn't necessarily indicate that it is better to use a higher capacity by default.

And your replies are not really answering why it's not necessary or desirable to couple the default capacity to the value of `k`.

Reply:
I think the main point of topk with sketch is that it has fixed memory assumption based on the capacity. If data becomes really large, the user can specify by the capacity how much memory he would like to use.

Reply:
So, since the memory consumption can be fixed, this is the argument against a dynamic growing capacity.

Reply:
Regarding the performance, it caught originally my attention based on the profiling and then found this curve https://datasketches.apache.org/docs/Frequency/FrequentItemsPerformance.html

Reply:
I think if we make a dynamically generated capacity, it would be better to calculate that from the table size to go with the optimum.

Reply:
> And your replies are not really answering why it's not necessary or desirable to couple the default capacity to the > value of k.

Afaik `k` is not the leading factor here. The main factor is number of processed items and capacity (=available memory) for each job.

Reply:
Anyway, what would be your suggestion here ? Leave it as it ? Just make it possible to add the capacity as a limit without the new defaults ?

Reply:
A query like:

```sql
select "cCode", topk("searchWord", 10) from uservisits group by 1 order by 1
```

With the uservisits dataset we have in crate-benchmarks and 2G heap will run
into a circuit breaker exception with the change here. Without the change it
works.

On the other hand, with the current default the accuracy looks rather bad

This is the real count:

```
cr> select "searchWord", count(*) from uservisits where "cCode" = 'AUT' group by 1 order by 2 desc limit 10;
+------------+----------+
| searchWord | count(*) |
+------------+----------+
| antiphoner |       53 |
| cathodes   |       53 |
| revoke     |       53 |
| unwaking   |       53 |
| Behistun   |       52 |
| neology    |       52 |
| dithymol   |       52 |
| ovaritides |       51 |
| pyronines  |       51 |
| palaeogene |       51 |
+------------+----------+
```

These are the approximates:

```
cr> select unnest(topk("searchWord", 8, 16)) from uservisits where "cCode" = 'AUT' ;
+--------------------------------------------------+
| unnest(topk("searchWord", 8, 16))                |
+--------------------------------------------------+
| {"frequency": 2713, "item": "piezometrical"}     |
| {"frequency": 2713, "item": "preignition"}       |
| {"frequency": 2713, "item": "sidebands"}         |
| {"frequency": 2713, "item": "tree-embowered"}    |
| {"frequency": 2713, "item": "VAP"}               |
| {"frequency": 2713, "item": "unwaking"}          |
| {"frequency": 2713, "item": "chronogrammatical"} |
+--------------------------------------------------+

cr> select unnest(topk("searchWord", 8, 2 ^ 8)) from uservisits where "cCode" = 'AUT' ;
+--------------------------------------------+
| unnest(topk("searchWord", 8, 256))         |
+--------------------------------------------+
| {"frequency": 185, "item": "inebriative"}  |
| {"frequency": 184, "item": "bradytely"}    |
| {"frequency": 184, "item": "ovaritides"}   |
| {"frequency": 184, "item": "Talegallinae"} |
| {"frequency": 184, "item": "matters"}      |
| {"frequency": 184, "item": "twice-worn"}   |
| {"frequency": 184, "item": "flypaper"}     |
| {"frequency": 184, "item": "refurnished"}  |
+--------------------------------------------+
SELECT 8 rows in set (0.006 sec)
cr> select unnest(topk("searchWord", 8, 2 ^ 9)) from uservisits where "cCode" = 'AUT' ;
+-------------------------------------------+
| unnest(topk("searchWord", 8, 512))        |
+-------------------------------------------+
| {"frequency": 96, "item": "dispending"}   |
| {"frequency": 96, "item": "ovaritides"}   |
| {"frequency": 96, "item": "metageometry"} |
| {"frequency": 95, "item": "flypaper"}     |
| {"frequency": 95, "item": "Teutolatry"}   |
| {"frequency": 95, "item": "carabaos"}     |
| {"frequency": 94, "item": "uther"}        |
| {"frequency": 94, "item": "meticulously"} |
+-------------------------------------------+
SELECT 8 rows in set (0.006 sec)
cr> select unnest(topk("searchWord", 8, 2 ^ 10)) from uservisits where "cCode" = 'AUT' ;
+-------------------------------------------+
| unnest(topk("searchWord", 8, 1024))       |
+-------------------------------------------+
| {"frequency": 65, "item": "metageometry"} |
| {"frequency": 65, "item": "Mantua"}       |
| {"frequency": 64, "item": "ovaritides"}   |
| {"frequency": 64, "item": "dithymol"}     |
| {"frequency": 64, "item": "oar-footed"}   |
| {"frequency": 63, "item": "antiphoner"}   |
| {"frequency": 63, "item": "dispending"}   |
| {"frequency": 63, "item": "darkens"}      |
+-------------------------------------------+
```


> I think if we make a dynamically generated capacity, it would be better to calculate that from the number of rows in the table.

Why? You might have intermediate operators, and number of rows != cardinality of the expression used for `topk`
Please state rationale and properties by which something is better if you make "would be better" statements.


---

So in short:

- Yes we should add a `capacity` parameter
- We should improve the docs to give the users some kind of info on what capacity yields what can of accuracy.

Reply:
> Yes we should add a capacity parameter

Do you have an opinion if the capacity parameter should be in the power of two or do prefer accepting any value and do internally rounding ?

>  We should improve the docs to give the users some kind of info on what capacity yields what can of accuracy.

What do you think about returning the an error/acuracy value from the sketch as part of the result ?

Reply:
> Do you have an opinion if the capacity parameter should be in the power of two or do prefer accepting any value and do internally rounding ?

Require power of two. Doing rounding would just obfuscate what's happening

> What do you think about returning the an error/acuracy value from the sketch as part of the result ?

That sounds nice. How would you represent that?

Reply:
> Please state rationale and properties by which something is better if you make "would be better" statements.
The relation between number of processed items, number of stored items and n max size (capacity) is described in this https://datasketches.apache.org/docs/Frequency/FrequentItemsErrorTable.html For example with a max capacity of 8,192 it is possible to detect the frequency of 1 million items if it's 854 or less. The sketch itself provides relative maximum error which we could return as part of the result e.g.:

Exact:
```
cr> select "searchWord", count(*) from uservisits  group by 1 order by 2 desc limit 10;
+----------------+----------+
| searchWord     | count(*) |
+----------------+----------+
| CTTN           |     4098 |
| german         |     4086 |
| tarlatans      |     4044 |
| uredo-fruit    |     4041 |
| magnetizations |     4033 |
| Shirleen       |     4032 |
| Downe          |     4031 |
| Biamonte       |     4027 |
| etagere        |     4017 |
| verdugoship    |     4009 |
+----------------+----------+
```
Approximation with no error:
```
cr> select unnest(topk("searchWord", 8, 2048)) from uservisits;
+-----------------------------------------------+
| unnest(topk("searchWord", 8, 2048))           |
+-----------------------------------------------+
| {"maximum_error": 0}                          |
| {"frequency": 4098, "item": "CTTN"}           |
| {"frequency": 4086, "item": "german"}         |
| {"frequency": 4044, "item": "tarlatans"}      |
| {"frequency": 4041, "item": "uredo-fruit"}    |
| {"frequency": 4033, "item": "magnetizations"} |
| {"frequency": 4032, "item": "Shirleen"}       |
| {"frequency": 4031, "item": "Downe"}          |
| {"frequency": 4027, "item": "Biamonte"}       |
+-----------------------------------------------+

```
Approximation with error:
```
cr> select unnest(topk("searchWord", 8, 1024)) from uservisits;
+----------------------------------------------+
| unnest(topk("searchWord", 8, 1024))          |
+----------------------------------------------+
| {"maximum_error": 6328}                      |
| {"frequency": 6340, "item": "thievingly"}    |
| {"frequency": 6340, "item": "Culpeper"}      |
| {"frequency": 6337, "item": "subacutely"}    |
| {"frequency": 6336, "item": "subglobosity"}  |
| {"frequency": 6335, "item": "underthrust"}   |
| {"frequency": 6335, "item": "unsaved"}       |
| {"frequency": 6335, "item": "nervelessness"} |
| {"frequency": 6335, "item": "Treharne"}      |
+----------------------------------------------+
```
The maximum relative error describes how much the frequency of an item can be off at maximum. So my suggestion would to provide some kind of metric in the docs how to choose a reasonable capacity and return the error.

Reply:
How about exposing all the info we have:

Turn the result from `list<map<...><` into an object with:

```
{
    "frequencies": [
        {
            "item": <expr>,
            "estimate": <item.estimate>,
            "lower_bound": <item.getlowerBound>,
            "upper_bound": <item.getupperbound>,
        },
        ....
    ],
    "maximum_error": <max_error>
}
```